### PR TITLE
fix(quantization): `nvfp4_quantize(backend='cuda')` silently corrupts scale factors when global_scale is not float32

### LINKED ISSUE #3398: mm_fp4 NVFP4 returns erratic M-dependent zeros on SM120: both b12x and cutlass backends affected in FlashInfer 0.6.12
### Summary

This may be related to #2577, but I am opening a separate issue because this is on FlashInfer 0.6.12 with the new `b12x` backend, and the failure is non-monotonic across M rather than a simple "all shapes fail" case.

On NVIDIA RTX PRO 6000 Blackwell (SM120, compute capability 12.0), `flashinfer.mm_fp4` returns all-zero outputs for specific M values. The pattern is erratic: M=1536/2048/4096 work, but M=3072 returns all zeros despite being between two working values.

Both `b12x` and `cutlass` show the same failure/work pattern in this M sweep, which suggests this may be a shared SM120 FP4 dispatch/tile-selection issue rather than a wrapper-only problem.

### Environment

- GPU: NVIDIA RTX PRO 6000 Blackwell Server Edition
- Compute capability: SM120 / CC 12.0
- Driver: 580.126.09
- CUDA toolkit: 12.8.93 system toolkit
- nvcc used for JIT: 13.2.78 from PyPI wheel `nvidia-cuda-nvcc==13.2.78`
- FlashInfer: 0.6.12
- PyTorch: 2.11.0+cu130
- OS: Linux 6.6.122 x86_64

### JIT workaround used

The default environment could not JIT-compile the SM120 FP4 kernels cleanly. To get the kernel to compile and launch, I used the following local workarounds:

1. Installed `nvidia-cuda-nvcc==13.2.78` because nvcc 12.8 does not accept `compute_120f`.
2. Symlinked `nvidia/cu13/lib/*` into `nvidia/cu13/lib64/` so the JIT linker could find CUDA runtime libraries.
3. Patched `flashinfer/compilation_context.py` locally so the CUDA 12.8 path does not emit an unsupported `compute_120f` target.

After these workarounds, JIT compilation succeeds and the kernels launch without crashing. The issue is silent incorrect output: many shapes return all zeros.

### Reproduction

```python
import os
os.environ["FLASHINFER_DISABLE_VERSION_CHECK"] = "1"

import torch
from flashinfer import mm_fp4, nvfp4_quantize

print("flashinfer:", __import__("flashinfer").__version__)
print("GPU:", torch.cuda.get_device_name())
print("capability:", torch.cuda.get_device_capability())

K, N = 2048, 8192  # Gemma-style MLP up-projection shape
torch.manual_seed(42)

for M in [256, 384, 512, 1024, 1280, 1536, 2048, 3072, 4096]:
    a = torch.randn(M, K, device="cuda", dtype=torch.bfloat16) * 5
    b = torch.randn(N, K, device="cuda", dtype=torch.bfloat16) * 5

    a_scale = (448.0 * 6.0) / a.abs().max()
    b_scale = (448.0 * 6.0) / b.abs().max()

    # Use nvfp4_quantize defaults, matching the run that produced the output below.
    a_fp4, a_sf = nvfp4_quantize(a, a_scale)
    b_fp4, b_sf = nvfp4_quantize(b, b_scale)

    alpha = torch.tensor(
        1.0 / (a_scale.item() * b_scale.item()),
        device="cuda",
        dtype=torch.float32,
    )

    for backend in ["b12x", "cutlass"]:
        kwargs = {"backend": backend}
        if backend == "b12x":
            kwargs["use_nvfp4"] = True

        out = mm_fp4(
            a_fp4,
            b_fp4.T,
            a_sf,
            b_sf.T,
            alpha,
            torch.bfloat16,
            **kwargs,
        )
        torch.cuda.synchronize()

        zero = bool((out == 0).all().item())
        max_abs = float(out.abs().max().item())
        print(f"M={M:5d}  {backend:8s}  zero={zero}  max={max_abs:.1f}")
```

### Actual output

```text
M=  256  b12x      zero=True   max=0.0
M=  256  cutlass   zero=True   max=0.0
M=  384  b12x      zero=True   max=0.0
M=  384  cutlass   zero=True   max=0.0
M=  512  b12x      zero=True   max=0.0
M=  512  cutlass   zero=True   max=0.0
M= 1024  b12x      zero=True   max=0.0
M= 1024  cutlass   zero=True   max=0.0
M= 1280  b12x      zero=True   max=0.0
M= 1280  cutlass   zero=True   max=0.0
M= 1536  b12x      zero=False  max=264.0
M= 1536  cutlass   zero=False  max=264.0
M= 2048  b12x      zero=False  max=231.0
M= 2048  cutlass   zero=False  max=231.0
M= 3072  b12x      zero=True   max=0.0
M= 3072  cutlass   zero=True   max=0.0
M= 4096  b12x      zero=False  max=276.0
M= 4096  cutlass   zero=False  max=276.0
```

### Expected behavior

All tested M values should produce non-zero output close to the BF16 reference matmul. M=3072 should not fail while both M=2048 and M=4096 work.

### Why this matters

Autoregressive decode commonly runs at small M values:

- single-sequence decode: M=1
- batch decode: M=batch size, often 1-256
- chunked prefill: M depends on chunk and batch shape

With this behavior, SM120 deployments cannot safely use `mm_fp4` for NVFP4 decode or mixed prefill/decode workloads, because some otherwise normal M values silently produce zeros. This forces users on RTX PRO 6000 / RTX 5090-class SM120 GPUs to fall back to Marlin W4A16 paths instead of native FP4 Tensor Core GEMM.

### Secondary observation

For a working case such as M=4096, I also saw relatively high error compared to BF16 matmul in my quick local check. I am not treating this as the primary bug because the scale choice in this synthetic repro may not match production calibration. The main bug here is the all-zero output for specific M values.

### Possible root cause

This looks like a shape-dependent tile selection or dispatch issue in the SM120 FP4 path. The non-monotonic behavior is the key fingerprint:

- M=1536 works
- M=2048 works
- M=3072 returns all zeros
- M=4096 works

The fact that `b12x` and `cutlass` show the same pattern suggests either shared kernel code, shared layout assumptions, or a shared dispatch/tile selection problem.

Happy to provide additional K/N sweeps, more M values, or rerun this in a clean environment if useful.

### PR DESCRIPTION
**Pull request by bkryu**
### Summary
Fixes #3398. `flashinfer.mm_fp4` / `nvfp4_quantize` (cuda backend) silently produced **all-zero or magnitude-wrong outputs** for certain batch sizes M when global scales are not provided as bfloat16. The root cause in the issue is that the CUDA quantize kernel reads the global scale as `float32`, but callers commonly pass a **bfloat16** global scale — e.g. `(448 * 6) / x.abs().max()` inherits `x`'s bf16  dtype. The kernel then misreads it byte-wise, and a dtype guard that would have caught this was commented out.

The PR fixes the issue by converting non-fp32 global scale factors to fp32 and adding checks. Also adds unit tests.

## 🔍 Related Issues

#3398

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Bug Fixes**
  * FP4 quantization now validates and normalizes the global scale so kernels always receive a float32 scale on the correct device, preventing mis-scaling or all-zero outputs when scale is bfloat16/float16 or on a different device.

* **Tests**
  * Added a regression test covering global-scale handling across bfloat16, float16, and float32.

### CODE REVIEW

File: flashinfer/quantization/fp4_quantization.py [hunk 1]
```diff
@@ -859,6 +859,10 @@ def fp4_quantize(
     assert input.shape[-1] % sf_vec_size == 0
     if enable_pdl is None:
         enable_pdl = device_support_pdl(input.device)
+    # Kernel reads global_scale as float32; normalize so a bf16/fp16 scale isn't
+    # misread byte-wise into corrupt scale factors (issue #3398). No-op if float32.
+    if global_scale is not None and global_scale.dtype != torch.float32:
+        global_scale = global_scale.to(torch.float32)
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/quantization/fp4_quantization.py]: Good catches on both — agreed, fixed in the latest commit.

**Review by aleozlx — APPROVED** [review]
